### PR TITLE
Kubelet: Add image cache.

### LINKED ISSUE #38373: Kubelet should not rely on the runtime to generate node status.
Currently, kubelet calls `ListImages` to get image list directly from the runtime (see [here](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet_node_status.go#L562) and [here](https://github.com/kubernetes/kubernetes/blob/1dfd64f4378ad9dd974bbfbef8e90127dce6aafe/pkg/kubelet/images/image_gc_manager.go#L150)).

We recently hit an issue https://github.com/docker/docker/issues/21215. Bulk image deletion/creation will cause `ListImages` to hang.
When that happens, kubelet has to wait until `ListImages` times out (2 min). If node status is updated by other components during that time (NPD, route controller, node controller etc.), kubelet will not be able to update status because of conflict.

Usually, when that happens, container should be able to work normally. However, pods will be evicted by node controller unnecessarily because kubelet can't update node status for a long time, and node controller thinks the node is `NotReady`.

Kubelet should not rely on the runtime to generate node status.

@yujuhong 
/cc @kubernetes/sig-node

### PR DESCRIPTION
**Pull request by Random-Liu**
Fixes #38373.

This should be patched into 1.5.1 to solve the customer issue.

@yujuhong 
/cc @kubernetes/sig-node

### CODE REVIEW

**Comment by kargakis** [comment]
Not sure how useful this can be for you guys but worth taking a look
@legionus @miminar @mfojtik @soltysh

**Comment by feiskyer** [comment]
> Fixes 38373.

nit: link to the issue #38373.

Jenkins unit/integration test is failed, forgotten to update the related tests?

File: pkg/kubelet/images/image_gc_manager.go [hunk 1]
```diff
@@ -142,16 +167,23 @@ func (im *realImageGCManager) Start() error {
 		}
 	}, 5*time.Minute, wait.NeverStop)
 
+	// Start a goroutine periodically updates image cache.
+	// TODO(random-liu): Merge this with the previous loop.
```
Comment (yujuhong) [inline pkg/kubelet/images/image_gc_manager.go]: I don't think we can easily merge the two loops. Not sure what this TODO is meant for.

**Review by yujuhong — COMMENTED** [review]
Yep, the unit tests needs fixing.

**Comment by ncdc** [comment]
cc @derekwaynecarr @smarterclayton for OpenShift backport consideration

File: pkg/kubelet/images/image_gc_manager.go (hunk 1 above)
Comment (Random-Liu) [inline pkg/kubelet/images/image_gc_manager.go]: `detectImages` also list image, in fact we can image list there. However, it doesn't list images as frequently as we need, so I add a TODO here to consider this later if needed. :)

**Comment by Random-Liu** [comment]
@yujuhong Fixed the unit test.

**Comment by smarterclayton** [comment]
Yes to backport in Kube 1.4 and openshift I think.  Status hangs are bad

**Pull request merged**